File: 24.Web-Sockets-And-Real-time-Communication-with-WebSockets/code/typescript/authentication_4.ts
import { WebSocketServer } from 'ws';
import * as http from 'http';

const server = http.createServer();
const wss = new WebSocketServer({ noServer: true });

async function verifyJWT(token: string): Promise<any> {
    // your auth logic (auth chapter)
    if (token === "valid") return { id: 1, name: "User" };
    throw new Error("Invalid token");
}

server.on('upgrade', async (request, socket, head) => {
    try {
        const url = new URL(request.url || '', `http://${request.headers.host}`);
        // AUTHENTICATE FIRST , before upgrading. Reject with a normal HTTP error.
        const token = url.searchParams.get("token") || ''; // or read a cookie / subprotocol
        
        const user = await verifyJWT(token);
        
        // Only now perform the upgrade; attach the identity to the connection.
        wss.handleUpgrade(request, socket, head, (ws) => {
            wss.emit('connection', ws, request, user);
        });
    } catch (err) {
        // 401, no upgrade
        socket.write('HTTP/1.1 401 Unauthorized\r\n\r\n');
        socket.destroy();
    }
});

wss.on('connection', (ws, request, user) => {
    // every message from this conn is now tied to `user`
    serve(ws, user);
});

function serve(ws: any, user: any) {
    ws.on('message', (msg: any) => {
        console.log(`Message from ${user.name}: ${msg}`);
    });
}

File: 24.Web-Sockets-And-Real-time-Communication-with-WebSockets/code/rust/authentication_4.rs
use axum::{
    extract::{Query, WebSocketUpgrade, ws::WebSocket},
    response::{IntoResponse, Response},
    http::StatusCode,
};
use serde::Deserialize;

#[derive(Deserialize)]
struct AuthQuery {
    token: Option<String>,
}

struct User {
    id: i32,
    name: String,
}

fn verify_jwt(token: &str) -> Result<User, ()> {
    // your auth logic (auth chapter)
    if token == "valid" {
        Ok(User { id: 1, name: "User".to_string() })
    } else {
        Err(())
    }
}

async fn ws_handler(
    ws: WebSocketUpgrade,
    // AUTHENTICATE FIRST , before upgrading. Reject with a normal HTTP error.
    query: Query<AuthQuery>, // or read a cookie / subprotocol
) -> Response {
    let token = query.token.clone().unwrap_or_default();
    
    let user = match verify_jwt(&token) {
        Ok(u) => u,
        Err(_) => return (StatusCode::UNAUTHORIZED, "unauthorized").into_response(), // 401, no upgrade
    };

    // Only now perform the upgrade; attach the identity to the connection.
    ws.on_upgrade(move |socket| serve(socket, user))
}

async fn serve(mut socket: WebSocket, user: User) {
    // every message from this conn is now tied to `user`
    while let Some(Ok(msg)) = socket.recv().await {
        println!("Message from {}: {:?}", user.name, msg);
    }
}
